prov/sockets: Fix compare buffer bug in fi_compare_atomic

shantonu opened the pull request:
When <code>FI_INJECT</code> is set as a flag in <code>fi_compare_atomic</code>, the compare data was not being copied to the internal buffer. This should fix #1959.

@jithinjosepkl
Signed-off-by: Shantonu Hossain <shantonu.hossain@intel.com>

jithinjosepkl commented:
@shantonu - Thanks for the patch. This patch writes the compare_buf to the TX ringbuffer. In addition to this, we need to read this data from the ringbuffer (rather than reading it directly from the compare-buffer) while sending out the atomic op control message.

shantonu commented:
Thanks! I have updated <code>sock_pe_new_tx_entry</code> to read from the ring buffer in case <code>FI_INJECT</code> flag is set.

jithinjosepkl commented on prov/sockets/src/sock_atomic.c:
```diff
@@ -161,24 +163,41 @@ ssize_t sock_ep_tx_atomic(struct fid_ep *ep,
 					  msg->msg_iov[i].count * datatype_sz);
 			src_len += (msg->msg_iov[i].count * datatype_sz);
 		}
+		for (i = 0; i < compare_count; i++) {
```
While processing the ringbuffer, the compare iovecs/data are expected after result iovecs. Doing it here will result in data corruption.

shantonu replied:
I have changed the processing part as well. I grouped the ringbuffer copy/read in the same place as otherwise another check is needed for compare data.

jithinjosepkl commented on prov/sockets/src/sock_atomic.c:
```diff
@@ -161,24 +163,41 @@ ssize_t sock_ep_tx_atomic(struct fid_ep *ep,
 			sock_tx_ctx_write(tx_ctx, &tx_iov, sizeof(tx_iov));
 			src_len += (tx_iov.ioc.count * datatype_sz);
 		}
+		for (i = 0; i < compare_count; i++) {
```
Same comment as above.

jithinjosepkl commented on prov/sockets/src/sock_progress.c:
```diff
@@ -2323,6 +2327,12 @@ static int sock_pe_new_tx_entry(struct sock_pe *pe, struct sock_tx_ctx *tx_ctx)
+				 	sizeof(pe_entry->pe.tx.tx_iov[i].cmp));
+				msg_hdr->msg_len += datatype_sz *
+					pe_entry->pe.tx.tx_iov[i].cmp.ioc.count;
+			}
```
Ok, noticed that the ringbuffer processing is updated as well. Now, compare-buffer is expected after src_buffer. 👍

jithinjosepkl replied:
Note that while sending out the atomic op request, the compare data is read from the iovecs. This should be updated to read from the inject buffer (+ offset).

jithinjosepkl replied:
Also, the src_iov_len in case of inject indicates the total data length (not the iov length). We can follow similar semantics for compare data as well.

shantonu replied:
ok, will update the cmp_iov_len to the data length for inject case.

shantonu commented:
Updated the PR with reading from ring buffer while sending out atomic op request in case FI_INJECT flag is set. Modified fabtests/rdm_atomic.c with FI_INJECT op_flags and verified the inject path.

jithinjosepkl commented on prov/sockets/src/sock_progress.c:
```diff
@@ -1826,25 +1826,22 @@ static int sock_pe_progress_tx_atomic(struct sock_pe *pe,
+					&pe_entry->pe.tx.inject[0] + pe_entry->pe.tx.tx_op.src_iov_len,
+					pe_entry->pe.tx.tx_op.atomic.cmp_iov_len, len))
+			return 0;
+		len += pe_entry->pe.tx.tx_op.atomic.cmp_iov_len;
```
This violates the wire protocol. Receiver expects the compare data before the src data. Please check sock_pe_process_rx_atomic().

shantonu replied:
Fixed the ordering.

jithinjosepkl commented on prov/sockets/src/sock_progress.c:
```diff
@@ -1826,25 +1826,31 @@ static int sock_pe_progress_tx_atomic(struct sock_pe *pe,
-				       pe_entry->pe.tx.tx_iov[i].cmp.ioc.count *
-				       datatype_sz, len))
+					&pe_entry->pe.tx.inject[0],
+					pe_entry->pe.tx.tx_op.atomic.cmp_iov_len, len))
```
@shantonu - For the inject case, you are writing cmp_data at tx.inject[0] + src_len. But, in here, it is read from tx.inject[0].

shantonu replied:
Oh, sorry! Updated the patch with correct offset.

jithinjosepkl commented on prov/sockets/src/sock_progress.c:
```diff
@@ -1826,25 +1826,31 @@ static int sock_pe_progress_tx_atomic(struct sock_pe *pe,
+		/* data */
 		if (sock_pe_send_field(pe_entry,
-				       &pe_entry->pe.tx.inject[0],
+				       &pe_entry->pe.tx.inject[0] + pe_entry->pe.tx.tx_op.atomic.cmp_iov_len,
```
Similar comment here, the src_data is written at tx.inject[0].

shantonu commented:
Modified fabtests/rdm_atomic.c and performed data verification on both inject and non-inject path for CSWAP.
